docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,74 @@
+- Status: In progress
+- Date: 2023-11-02
+- Author: [@gela](https://github.com/gela)
+- RFC PR: TBD
+- Implementation GitHub issue: TBD.
+
+
+## Summary
+
+The existing `ConnectionManager` class will be renamed to `ConnectionDiagnostics` and will be responsible for handling all connectivity related checks once during VPN activation (`Controller::StateConnecting`) and then periodically once the VPN is already activated (`Controller::StateOn`); the controller will be the entity which manages the controller states that the client goes through such as `connecting` and `confirming` as we activate and deactivate the VPN to reflect those to the user what is happening behind the scene.
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks consistency for users. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a "background error occurred" banner or lose signal without further information.
+
+The objective with the Connection Diagnostics is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, this document proposes encapsulating this new functionality within the Connection Diagnostics. This object will interact with both the controller and connection health, serving as a dedicated entity for identifying connection problems and correctly communicating status changes to the UI, ultimately enhancing the user experience.
+
+## Proposed Solution
+
+The Connection Diagnostics will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionDiagnostics::State`. The Connection Diagnostics will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful. 
+
+In the event of a probe failure, the Connection Diagnostics will emit a signal to the frontend. This will trigger a modal to be displayed to the user, providing an explanation of the issue and suggesting potential solutions. Upon encountering a failed probe, the progression through the remaining probes is immediately halted and the Connection Diagnostics will enter the Idle state.
```

Comment:
What if there is instability in the network and all ConnectionDiagnostics probes succeed? Do we leave the user hanging or is there a catch all error modal?

Reply:
Also what if ConnectionHealth recovers while ConnectionDiagnostics is still running probes? Does it abort? In that same vein, what if ConnectionHealth recovers while the modal is showing to users? Does it just disappear?

Reply:
Just noticed this is the same question Matt just asked, :shrug:

Reply:
If all probes pass but VPN is still in No Signal (aka we failed to pinpoint the exact issue) then we won't show any modals to the user. At that point we'll just leave it in No Signal, not sure a catch-all modal would offer additional guidance

Reply:
I'm fine with that, maybe call it out in the doc.

Reply:
@brizental Added to the doc about 20 seconds before your commend 😄

```
- If all probes pass but the stability is still `NoSignal`:
  - Something else has gone wrong and are not able to determine the cause. VPN continues to stay in No Signal but we will not show any further modals to the user
  - This will be recorded in the logs for debugging purposes
  ```